### ISSUE TITLE
LuaUnsyncedRead::GetSelectedUnitsCounts creates an obscene amount of garbage

### ISSUE THREAD

**Issue by Beherith** [beyond-all-reason/spring#1220]
It creates 266 kilobytes of garbage every time it is called, which is astonishingly high. The numbers don't even add up correctly, as its creating a table with 1300 elements in the hash part only:

`lua_createtable(L, 0, gsucCountMap.size());`

```
/*** Get an aggregate count of selected units per unitDefID
 *
 * @function Spring.GetSelectedUnitsCounts
 *
 * @treturn {[number]=number,...} unitsCounts where keys are unitDefIDs and values are counts
 * @treturn n the number of unitDefIDs
 */
int LuaUnsyncedRead::GetSelectedUnitsCounts(lua_State* L)
{
	gsucCountMap.clear();
	gsucCountMap.resize(unitDefHandler->NumUnitDefs() + 1, {0, 0});

	int numDefKeys = 0;

	// tally the types
	for (const int unitID: selectedUnitsHandler.selectedUnits) {
		const CUnit* unit = unitHandler.GetUnit(unitID);
		const UnitDef* unitDef = unit->unitDef;

		gsucCountMap[unitDef->id].first = unitDef->id;
		gsucCountMap[unitDef->id].second += 1;
	}

	// { [number unitDefID] = number count, ... }
	LOG("gsucCountMap.size() = %d", gsucCountMap.size());
	lua_createtable(L, 0, gsucCountMap.size());

	for (const std::pair<int, int>& p: gsucCountMap) {
		if (p.second == 0)
			continue;

		lua_pushnumber(L, p.second); // push the UnitDef unit count (value)
		lua_rawseti(L, -2, p.first); // push the UnitDef index (key)

		numDefKeys += 1;
	}

	lua_pushnumber(L, numDefKeys);

	return 2;
}
```

**Comment by Beherith** [comment]
Is this a valid fix? 
```

/*** Get an aggregate count of selected units per unitDefID
 *
 * @function Spring.GetSelectedUnitsCounts
 *
 * @treturn {[number]=number,...} unitsCounts where keys are unitDefIDs and values are counts
 * @treturn n the number of unitDefIDs
 */
int LuaUnsyncedRead::GetSelectedUnitsCounts(lua_State* L)
{
	gsucCountMap.clear();
	gsucCountMap.resize(unitDefHandler->NumUnitDefs() + 1, {0, 0});

	int numDefKeys = 0;

	int numSelectedDefs = 0;

	// tally the types
	for (const int unitID: selectedUnitsHandler.selectedUnits) {
		const CUnit* unit = unitHandler.GetUnit(unitID);
		const UnitDef* unitDef = unit->unitDef;

		gsucCountMap[unitDef->id].first = unitDef->id;
		gsucCountMap[unitDef->id].second += 1;
		if (gsucCountMap[unitDef->id].second == 1) numSelectedDefs++;
	}

	// { [number unitDefID] = number count, ... }
	lua_createtable(L, 0, numSelectedDefs); // FIXME 

	for (const std::pair<int, int>& p: gsucCountMap) {
		if (p.second == 0)
			continue;

		lua_pushnumber(L, p.second); // push the UnitDef unit count (value)
		lua_rawseti(L, -2, p.first); // push the UnitDef index (key)

		numDefKeys += 1;
	}

	lua_pushnumber(L, numDefKeys);

	return 2;
}
```

**Comment by Beherith** [comment]
Or should I balance the cost of doing 
`if (gsucCountMap[unitDef->id].second == 1) numSelectedDefs++; `
vs the fixed cost of iterating over `gsucCountMap` to count non-zeros?

**Comment by Beherith** [comment]
This bug also affects all 4 cases of `*map.resize(unitDefHandler->NumUnitDefs()...);`

**Comment by sprunk** [comment]
I think the fix looks like this 8ab0b92350b14961fee0848e0ab5fc13122e3a0c

**Comment by Beherith** [comment]
@sprunk 
I love this test, its very elegant:
```
if (!counts[unitDef->id]++)
			numDefKeys++;
```

But I think there is a minor mistake in your commit:
Originally, the hash part is allocated too big:
`lua_createtable(L, 0, gsucCountMap.size());`

In you fix, you are allocating the array part, and then inserting non-consecutive integers:
`lua_createtable(L, numDefKeys, 0);`

Could you also apply your change to:
`int LuaUnsyncedRead::GetGroupUnitsSorted(lua_State* L)`

`int LuaUnsyncedRead::GetGroupUnitsCounts(lua_State* L)`

`int LuaUnsyncedRead::GetSelectedUnitsSorted(lua_State* L)`
